lib/much-stub.rb

```diff
@@ -1,4 +1,207 @@
 require "much-stub/version"
 
 module MuchStub
+
+  def self.stubs
+    @stubs ||= {}
+  end
+
+  def self.stub(obj, meth, &block)
+    (self.stubs[MuchStub::Stub.key(obj, meth)] ||= begin
+      MuchStub::Stub.new(obj, meth, caller_locations)
+    end).tap{ |s| s.do = block }
+  end
+
+  def self.unstub(obj, meth)
+    (self.stubs.delete(MuchStub::Stub.key(obj, meth)) || MuchStub::NullStub.new).teardown
+  end
+
+  def self.unstub!
+    self.stubs.keys.each{ |key| self.stubs.delete(key).teardown }
+  end
+
+  def self.stub_send(obj, meth, *args, &block)
+    orig_caller = caller_locations
+    stub = self.stubs.fetch(MuchStub::Stub.key(obj, meth)) do
+      raise NotStubbedError, "`#{meth}` not stubbed.", orig_caller.map(&:to_s)
+    end
+    stub.call_method(args, &block)
+  end
+
+  class Stub
+
+    def self.key(object, method_name)
+      "--#{object.object_id}--#{method_name}--"
+    end
+
+    attr_reader :method_name, :name, :ivar_name, :do
+
+    def initialize(object, method_name, orig_caller = nil, &block)
+      orig_caller ||= caller_locations
+      @metaclass   = class << object; self; end
+      @method_name = method_name.to_s
+      @name        = "__muchstub_stub__#{object.object_id}_#{@method_name}"
+      @ivar_name   = "@__muchstub_stub_#{object.object_id}_" \
+                     "#{@method_name.to_sym.object_id}"
+
+      setup(object, orig_caller)
+
+      @do     = block
+      @lookup = {}
+    end
+
+    def do=(block)
+      @do = block || @do
+    end
+
+    def call_method(args, &block)
+      @method.call(*args, &block)
+    end
+
+    def call(args, orig_caller = nil, &block)
+      orig_caller ||= caller_locations
+      unless arity_matches?(args)
+        msg = "arity mismatch on `#{@method_name}`: " \
+              "expected #{number_of_args(@method.arity)}, " \
+              "called with #{args.size}"
+        raise StubArityError, msg, orig_caller.map(&:to_s)
+      end
+      lookup(args, orig_caller).call(*args, &block)
+    rescue NotStubbedError => exception
+      @lookup.rehash
+      lookup(args, orig_caller).call(*args, &block)
+    end
+
+    def with(*args, &block)
+      orig_caller = caller_locations
```

Comment:
@kellyredding - Do we need a ruby 1.8.7 polyfill for this? Similar to what you did here: https://github.com/redding/assert/pull/286

Reply:
@jcredding great catch - thanks!